pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/internal/infer/Infer.java

```diff
@@ -561,52 +563,76 @@ private JMethodSig instantiateImpl(JMethodSig m, MethodCallSite site, MethodReso
             addArgsConstraints(infCtx, m, site, phase); // c
             infCtx.incorporate(); // b2
 
-            if (phase.isInvocation()) {
-
-                boolean shouldPropagate = shouldPropagateOutwards(m.getReturnType(), site, infCtx);
-
-                //propagate outwards if needed
-                if (shouldPropagate) {
-                    // propagate inference context outwards and exit
-                    // the outer context will solve the variables and call listeners
-                    // of this context
-                    LOG.propagateAndAbort(infCtx, site.getOuterCtx());
-                    infCtx.duplicateInto(site.getOuterCtx());
-                    return infCtx.mapToIVars(m);
-                }
-            }
-
-            // this may throw for incompatible bounds
-            boolean isDone = infCtx.solve(/*onlyBoundedVars:*/isPreJava8());
-
-            if (isPreJava8() && !isDone) {
-                // this means we're not in an invocation context,
-                // if we are, we must ignore it in java 7
-                if (site.getOuterCtx().isEmpty()) {
-                    // Then add the return contraints late
-                    // Java 7 only uses the context type if the arguments are not enough
-                    // https://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.12.2.8
-                    m = doReturnChecksAndChangeReturnType(m, site, infCtx);
+            // TODO : check why we allow some inferences to skip the invocation phase… maybe an optimization?
```

Comment:
@oowekyala I'd love your insights on this… I couldn't find in the JLS anything referring to this, so I assumed it was you being smart about when we actually needed another pass or not (given que old code committed changes in all phases).

If that's the case, this change should keep the behavior

Reply:
So this is in fact an optimization but it is also very natural. Just like skipping type inference if a method is not generic.

In [JLS 18](https://docs.oracle.com/javase/specs/jls/se17/html/jls-18.html#jls-18.5.2) when they introduce invocation type inference, they mention that it is like regular invocation with possibly additional constraints. We skip invocation when we determine that there can be no additional constraints, so that the result of another round of invocation inference would be the same as the current result we have (because inference with the same constraints gives the same result).

I will try to review this PR shortly and understand better what's going on...